### PR TITLE
Add overview page for the portability framework

### PR DESCRIPTION
**Pull request by herohde**
Manually adding a screenshot, because the automatic staging seems to be broken:

preview

### CODE REVIEW

**Comment by kennknowles** [comment]
retest this please

File: src/_includes/header.html [hunk 1]
```diff
@@ -84,6 +84,7 @@
             <li role="separator" class="divider"></li>
             <li class="dropdown-header">Technical References</li>
             <li><a href="{{ site.baseurl }}/contribute/design-principles/">Design Principles</a></li>
+	    <li><a href="{{ site.baseurl }}/contribute/portability/">Portability Framework</a></li>
```
Comment (kennknowles) [inline src/_includes/header.html]: My reading of the Jenkins log is that the job definition is fixed, but that the `site.baseurl` is set to http://beam.apache.org during link validation, rather than the actual base of the test run. So I think to make mergebot happy we can do this in a follow-up PR, or we have to manually merge, ignoring mergebot's complaints.

**Comment by kennknowles** [comment]
R: @melap

**Comment by kennknowles** [comment]
Jenkins didn't comment, but it is staged at http://apache-beam-website-pull-requests.storage.googleapis.com/340/index.html

File: src/_includes/header.html (hunk 1 above)
Comment (herohde) [reply src/_includes/header.html]: Ok. I'm just doing what the other pages do here.

Comment (kennknowles) [reply src/_includes/header.html]: Oh, yes, the change is right. The staged menu looks good.

File: src/contribute/portability.md [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+# Portability framework
+
+* TOC
+{:toc}
+
+## Overview
+
+Interoperability between SDKs and runners is a key aspect of Apache
+Beam. So far, however, the reality is that most runners support the
+Java SDK only, because each SDK-runner combination require non-trivial
```
Comment (melap) [inline src/contribute/portability.md]: requires

File: src/contribute/portability.md [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+ * **Job submission and management**: The _Runner API_ defines a
+   language-neutral pipeline representation with transformations
+   specifying the execution environment as a docker container
+   image. The latter both allows the execution side to set up the
+   right environment as well as opens the door for custom containers
+   and cross-environment pipelines. The _Job API_ allows pipeline
+   execution and configuration to be managed uniformly.
+
+ * **Job execution**: The _Fn API_ defines execution-time binary
```
Comment (melap) [inline src/contribute/portability.md]: missing an -> "defines an"

File: src/contribute/portability.md [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+   contract between the SDK harness and the runner for how execution
```
Comment (melap) [inline src/contribute/portability.md]: for how -> that describes how

File: src/contribute/portability.md [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+ * [Cross
+   language](https://s.apache.org/beam-mixed-language-pipelines). Options
+   and tradeoffs for how to handle various kinds of
+   multi-language/multi-SDK pipelines.
+
+## Development
+
+The portability framework is a substantial effort that touches every
+Beam component. Other than the sheer magnitude, a major challenge is
```
Comment (melap) [inline src/contribute/portability.md]: suggestion for slight reword of the first part of this sentence:
In addition to the sheer magnitude, a major challenge is engineering an interop layer that does not significantly compromise performance due ...

File: src/contribute/portability.md [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+engineering an interop layer that is not significantly compromising
+performance due to the additional serialization overhead of a
+language-neutral protocol.
+
+### Roadmap
+
+The proposed project phases are roughly as follows and are not
+strictly sequential, not least because various components will likely
```
Comment (melap) [inline src/contribute/portability.md]: not least because -> as

File: src/contribute/portability.md [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+   should work.
+
+### Issues
+
+The portability effort touches every component, so the "portability"
+label is used to identify all portability-related issues. Pure
+design or proto definitions should use the "beam-model" component. A
+common pattern for new portability features is that the overall
+feature is in "beam-model" with subtasks for each SDK and runner in
+their respective
```
Comment (melap) [inline src/contribute/portability.md]: not sure if this sentence is missing the end or the JIRA down below is the end. JIRA is on a separate line so a bit off either way. that JIRA link does not work for me -- it says the filter does not exist or is private, even when logged in.

File: src/contribute/portability.md [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+
+**JIRA:** [query](https://issues.apache.org/jira/issues/?filter=12341925)
+
+### Status
+
+MVP in progress. No SDK or runner yet supports the full portability
+API, but once that happens a more detailed progress table will be
```
Comment (melap) [inline src/contribute/portability.md]: recommend moving the "yet" to after "portability API"

File: src/contribute/portability.md [hunk 9]
```diff
@@ -0,0 +1,164 @@
+---
+layout: default
+title: "Portability framework"
```
Comment (melap) [inline src/contribute/portability.md]: Framework (right now we have all page titles as title case (cap every word))

File: src/contribute/portability.md [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+permalink: /contribute/portability/
+redirect_from: /portability/
+---
+
+# Portability framework
```
Comment (melap) [inline src/contribute/portability.md]: add caps: Framework

**Comment by herohde** [comment]
Thanks @melap and @kennknowles! Fixed the text as per the suggestions.

File: src/contribute/portability.md [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+ * **Job submission and management**: The _Runner API_ defines a
+   language-neutral pipeline representation with transformations
+   specifying the execution environment as a docker container
+   image. The latter both allows the execution side to set up the
+   right environment as well as opens the door for custom containers
+   and cross-environment pipelines. The _Job API_ allows pipeline
+   execution and configuration to be managed uniformly.
+
+ * **Job execution**: The _Fn API_ defines an execution-time binary
+   contract between the SDK harness and the runner that describes how
```
Comment (iemejia) [inline src/contribute/portability.md]: Maybe introducing before what is or what is the goal of the SDK harness will make this paragraph clearer.

Comment (herohde) [reply src/contribute/portability.md]: Good idea. Thanks!

**Comment by herohde** [comment]
Thanks @iemejia. PTAL

**Comment by iemejia** [comment]
Excellent 👍

**Comment by herohde** [comment]
Thanks!

**Comment by melap** [comment]
@asfgit merge

**Comment by kennknowles** [comment]
@asfgit merge

File: src/_includes/section-menu/contribute.html [hunk 12]
```diff
@@ -16,6 +16,7 @@
 
   <ul class="section-nav-list">
     <li><a href="{{ site.baseurl }}/contribute/design-principles/">Design principles</a></li>
+    <li><a href="{{ site.baseurl }}/contribute/portability/">Portability Framework</a></li>
```
Comment (melap) [inline src/_includes/section-menu/contribute.html]: will need to remove this entry as a workaround for the issue when testing for new files, and put in a separate PR after this one is in. there's a JIRA for me to fix the root cause of that (BEAM-3144)

**Comment by kennknowles** [comment]
And actually before we have mergbot merge it, I think the commits need squashing, no? Not important, but it will make history more legible.

File: src/_includes/section-menu/contribute.html (hunk 12 above)
Comment (herohde) [reply src/_includes/section-menu/contribute.html]: Sorry, but are you asking me to do that or are you doing it?

Comment (herohde) [reply src/_includes/section-menu/contribute.html]: Done

**Comment by herohde** [comment]
Squashed the commits as well. @melap PTAL

**Comment by melap** [comment]
@asfgit merge

**Comment by asfgit** [comment]
MergeBot: Merge: PR 340 in state failure with description Timed out trying to find verification job. Check Jenkins (https://builds.apache.org/job/beam_PreCommit_Website_Merge/) to ensure job is configured correctly.. URL: http://mergebot-vm.apache.org/beam-website/340.

**Comment by asfgit** [comment]
Error: Timed out trying to find verification job. Check Jenkins (https://builds.apache.org/job/beam_PreCommit_Website_Merge/) to ensure job is configured correctly.

**Comment by melap** [comment]
@asfgit merge

**Comment by asfgit** [comment]
Error: PR failed in verification; check the Jenkins job for more information.

**Comment by melap** [comment]
retest this please

**Comment by herohde** [comment]
Removed the redirect

**Comment by melap** [comment]
@asfgit merge
